### ISSUE TITLE
remove CapsLock metakey

### ISSUE THREAD

**Issue by gingerr** [lwouis/alt-tab-macos#61]
The CapsLock one is a special one. It doesn't trigger an keyUp event after its depressed so it stays active which is an issue since to hide the UI you have to actively press CapsLock again. Is this by design or should it be removed as an metakey option?

**Comment by lwouis** [comment]
You're right, it doesn't work smoothly. I think `fn` is also not working as well.

I just checked HyperSwitch and WindowSwitcher, and both only list command/control/option. I will keep the same list and remove fn and capslock. This can be revisited later of course, but for now I think better be safe and have it working

**Comment by gingerr** [comment]
Great, that eliminates one problematic edge case to handle. 

Fn worked at my MacBook Pro 2015 keyboard but did not work on my pc keyboard (it seems like it’s simply not possible on macOS to watch that one event from non-apple keyboards) so that’s two edge cases eliminated then :)

Btw. did you get notifications for the pull requests?

**Comment by lwouis** [comment]
> Fn worked at my MacBook Pro 2015 keyboard but did not work on my pc keyboard (it seems like it’s simply not possible on macOS to watch that one event from non-apple keyboards) so that’s two edge cases eliminated then :)

I was looking into #17 and after digging deep into the history of OS APIs, and discussions online, I came to the conclusion that we need to use a library for these shortcuts. The one I want to try first is [ShortcutRecorder](https://github.com/Kentzo/ShortcutRecorder). I will help with #17, but also with #49 and also this ticket.

It will let users just press the shortcut they want, and have it work somehow, using very old Carbon APIs or even private APIs under the scene who knows. Essential part is that is will *work*. 

> Btw. did you get notifications for the pull requests?

Oh god no! I see there are 4 PRs now; 3 from you! I'm gonna find how to enable these notifications, and also check out the PRs!

**Comment by gingerr** [comment]
That shortcutrecorder looks interesting. Gonna take a look inside it’s source code after work out of curiosity.

I was not sure if you got notifications cause you answered this issue very fast but there was no communication in the pull requests.

Thanks 🤘🏻

**Comment by lwouis** [comment]
I have a first implementation of ShortcutRecorder working locally. Unfortunately, it [doesn't seem to support recording only a modifier key](https://github.com/Kentzo/ShortcutRecorder/issues/114).

@gingerr btw what kind of UI do you think would be best in the end? I'm hesitating again between every shortcut being a different preference (to accomodate users who prefer to use middle row shortcut for instance, or in general more flexibility to the user), or some kind of simplified version like just pick a meta key and the other keys are fixed (alt, shift, arrows, etc).

**Comment by gingerr** [comment]
@lwouis Damn, I was afraid of how ShortcutRecorder would handle this applications requirements since AltTab is/needs? to split up the recorded combination in a modifier (hold) and regular key (repeated use). Wondering also about the required logic when the user records a combination a three or more keys (differentiation between alt 'key' and tab 'key' part).

Btw did you check if it were able to catch/execute on pressing the recorded combination inside a native macOS password field? https://github.com/lwouis/alt-tab-macos/issues/17 That would be a win.

I would prefer the latter simplified version (simple code, simple menu).

In my opinion it would be nice and self explanatory if the preferences dialog would offer to specify the keys as simple as:
- Meta key: drop down as it is (maybe rename the label to Alt)
- Tab: drop down with e.g.:
  - Tab
  - the key left to 1
  - maybe the key between left shift and z/y (not sure if all keyboard layouts have that)
  - maybe something for left handers but I don't know how they alt tab usually on a mouse/keyboard setup

This keeps it stupid simple, does not need additional checks if the user records some key or key combination which is reserved by the application for additional functionality (left shift, arrow keys, ...), does not need to check if the user does record a two modifier keys combination (without regular key), does not need to check a special key like caps lock is involved, ...

Then to add additional keys as the application userbase requests and possibly to make it more powerful/advanced/complicated once the other killer features are solved (in my opinion: minimised windows, full screen windows/other spaces windows, slow showUI performance). Iam afraid my Swift/macOS knowledge is lacking here greatly. Currently aim practicing/trying to implement issue https://github.com/lwouis/alt-tab-macos/issues/28 

Iam not familiar with 'middle row shortcut', what do you mean by that?

I could implement this small change (simplified version) with my limited knowledge swiftly. There is some investigation required how to render the localised key icons in the dropdown options for the tab key (if it allows different choices) and to check if the keycodes differ with different layouts or if its handled already by the device independent keycodes apple provides (kVK_...) or to use some library for that.

**Comment by lwouis** [comment]
> Btw did you check if it were able to catch/execute on pressing the recorded combination inside a native macOS password field? #17 That would be a win.

I just checked and it does! I think they use different, no longer documented Carbon APIs that have basically more options available and more power.

Btw I think these 3 tickets are related: #17, #49, #50

> [you proposing drop-downs for modifier and drop-downs for key]

You're basically suggesting what HyperSwitch is currently doing. It is very simple, I agree. That being said there are some issues I see with their approach is:

* It's limited (what if I want to use `command` + `E`? Why can't I use it? People have different preferences)
* It can clash. See what I'm able to do here:
![image](https://user-images.githubusercontent.com/106195/67837064-8a134a00-fb31-11e9-8a5f-c919e687f610.png)

> does not need additional checks if the user records some key or key combination which is reserved by the application for additional functionality

For that, ShortcutRecorder has validation facilities so you can tell if a shortcut is already taken by the OS or another app. We can display that to the user so that's ok.

> (left shift, arrow keys, ...)

For these, we could imagine a UI like this:

* `Shortcut to activate the app`: `shortcut recorder from SR` (which has alt+tab on first launch)
-> At this point we will use the meta key(s) from that shortcut for the following preferences:
* `Key to go to next window`: `key recorder from SR` (which doesn't let you use a modifier, only a key)
* `Key to go to previous window`: `key recorder from SR` (which doesn't let you use a modifier, only a key). Here we give up `shift`-`tab` though.
* `Key to hide the app`: `key recorder from SR` (which doesn't let you use a modifier, only a key)

Between your proposal and my proposal, the trade-off is simplicity vs flexibility. Not sure which one is useful to more user. Of course, my approach *requires* SR, so it may be done later as you suggest since yours is almost already done now with current tech.

> Iam not familiar with 'middle row shortcut', what do you mean by that?

In the vim/dworak communities, there is this idea of trying to have your fingers spend time in the rest row (with characters `asdfghjkl;'` on US QWERTY). Using the `escape` key in these communities for example is seen as undesirable, and lots of people remap `escape` which is used in vim all the time to change modes to `caps-lock` which most programmers only hit by accident (I have disabled caps-lock as a personal anecdote) but which has a nice position on the keyboard.

My point is that alt-tab was born of Windows making it a convention through their legacy, but we could open the doors to new layouts. I spent some time on Linux with i3, and even now on my mac I'm using some light form of window-manager, and I'm trying to have a set of shortcut to do common stuff like I had on i3:

* maximize a window
* make window 50% screen with and put on left/right (to do side-by-side)
* switch focus from left/right window with keyboard (during side-by-side work)
* focus another windows (this is where AltTab comes in)

Personally I think meta+tab is nice, but I'm thinking other people may want to do it differently.

> There is some investigation required how to render the localised key icons in the dropdown options

I'm gonna stop you right there lol See this is exactly why I believe using a library is critical. I found some tickets in SR repos which showed me the incredible complexity of localizing shortcuts. There are so many hardware keyboards from so many country, shortcuts can be right-to-left, etc, etc. It's basically madness to imagine covering that scope in AltTab. Akin to implementing a Calendar library from scratch ;p Let's use a lib and stop worrying about capturing keys and displaying them back to the user!

**Comment by gingerr** [comment]
> I just checked and it does! I think they use different, no longer documented Carbon APIs that have basically more options available and more power.

Thats great!

> Btw I think these 3 tickets are related: #17, #49, #50

Yeah, #17 was the reason i was asking here. Good reference. #50 sounds like a good replacement and enhancement for the native cmd + (key above tab) feature (which does not work with minimized and fullscreen applications if i remember correctly, we would rock productivity if we could implement this).

> It's limited (what if I want to use command + E? Why can't I use it? People have different preferences)

Good point.

> It can clash. See what I'm able to do here

Right, we will need some validation logic there.

> For that, ShortcutRecorder has validation facilities so you can tell if a shortcut is already taken by the OS or another app. We can display that to the user so that's ok.

Oh, thats sounds great. Don't get me wrong, I'm not at all against usage of libraries or frameworks.

**Your ShortcutRecorder logic suggestion**

> Shortcut to activate the app: shortcut recorder from SR (which has alt+tab on first launch)
-> At this point we will use the meta key(s) from that shortcut for the following preferences:

Good idea to split the recorded shortcut up. So you would create two ShortcutRecorder bindings out of the one the user entered, i see.

> Key to go to next window: key recorder from SR (which doesn't let you use a modifier, only a key)

Isnt that the recorded TAB key from the previous recorded shortcut?

> Key to go to previous window: key recorder from SR (which doesn't let you use a modifier, only a key). Here we give up shift-tab though.

In my opinion a user would need to choose here between a modifier or a key. The modifier would need to be used in combination with `Shortcut to activate the app`. Or the user would record a regular key, e.g. the key above TAB to get macOS native reverse tab switching.

> Key to hide the app: key recorder from SR (which doesn't let you use a modifier, only a key)

Wait, wouldnt this make the app be a toggle requiring another keypress instead of de-pressing the metakey?

> Between your proposal and my proposal, the trade-off is simplicity vs flexibility. Not sure which one is useful to more users. Of course, my approach requires SR, so it may be done later as you suggest since yours is almost already done now with current tech.

What does SR stand for (not familiar with that one, time for software development)? Regarding the choice to be made here: Your proposal is of course much better and powerful and will be implemented sooner or later anyway. The question or lets say choice or possibility is to do it now or later on.

> In the vim/dworak communities, there is this idea of trying to have your fingers spend time in the rest row (with characters asdfghjkl;' on US QWERTY). Using the escape key in these communities for example is seen as undesirable, and lots of people remap escape which is used in vim all the time to change modes to caps-lock which most programmers only hit by accident (I have disabled caps-lock as a personal anecdote) but which has a nice position on the keyboard.

I see, thanks for elaborating that one to me. I never gave vim a fair chance and usually have to try to remember how to quit it if i land in vim land by accident 😄  

I had looked into dworak and it sounds like a great layout but its already problematic for me at work to switch from US to DE keyboard layout. I switched to the US layout some time ago cause all the special characters are placed way better for programming (obviously) but when I am assisting/reviewing code at systems of my colleagues (most use a german keyboard layout) i make many typing mistakes.

Which keyboard layout are you using?

> My point is that alt-tab was born out of Windows making it a convention through their legacy, but we could open the doors to new layouts. I spent some time on Linux with i3, and even now on my mac I'm using some light form of window-manager, and I have to a set of shortcuts to do common stuff like I had on i3:
> - maximize a window
> - make window 50% screen with and put on left/right (to do side-by-side)
> - switch focus from left/right window with keyboard (during side-by-side work)
> - focus another windows (this is where AltTab comes in)
>
> Personally I think meta+tab is nice, but I'm thinking other people may want to do it differently.

Offering choices and additional features is great. Point taken. I would love if we could offer that shortcuts (in my case especially window splitting and maximize; currently i use BetterTouchTool for that).

> I'm gonna stop you right there lol See this is exactly why I believe using a library is critical. I found some tickets in SR repo which showed me the incredible complexity of localizing shortcuts. There are so many hardware keyboards from so many country, shortcuts can be right-to-left, etc, etc. It's basically madness to imagine covering that scope in AltTab. Akin to implementing a Calendar library from scratch ;p Let's use a lib and stop worrying about capturing keys and displaying them back to the user!

Hehe, thanks for that link. Was afraid this localization would be a pain in the ass. I am all in for libraries / frameworks for already solved problems 👍

**Comment by lwouis** [comment]
> The question or lets say choice or possibility is to do it now or later on

Let's open a ticket for better shortcut customization, and see when this should be done compared to other tasks. I think it will end up being low priority

> I would love if we could offer that shortcuts

This is included in #9 which is a high priority for me personally as well

> Which keyboard layout are you using?

I've had the same exact experience as you (everyone around me using French layout). There is no clear winner here. I personally moved to US layout a few years ago because they have the largest user base i care about (Chinese keyboard may soon/already have the largest user base, but their contributions and discussions are gated by language, culture, and the Great Firewall).

I moved to Tokyo a year ago so my choice/investment in US layout paid off, but there are still bad aspects like when i left i couldn't easily sell my macbook since normal people expect FR layout
